src/Pos/Launcher/Resource.hs

```diff
@@ -0,0 +1,380 @@
+loggerBracket lp = bracket_ (setupLoggers lp) releaseAllHandlers
+
+----------------------------------------------------------------------------
+-- NodeContext
+----------------------------------------------------------------------------
+
+allocateNodeContext ::
+       forall ssc m.
+       ( SscConstraint ssc
+       , SecurityWorkersClass ssc
+       , MonadIO m
+       , WithLogger m
+       , MonadMockable m
+       , MonadMask m
+       , MonadBase IO m
+       , MonadBaseControl IO m
+       , MonadDBRead m
+       )
+    => NodeParams
+    -> SscParams ssc
+    -> m $ NodeContext ssc
+allocateNodeContext np@NodeParams {..} sscnp = do
+    ncLoggerConfig <- getRealLoggerConfig $ bpLoggingParams npBaseParams
+    setupLoggers $ bpLoggingParams npBaseParams
+    ncBlkSemaphore <- BlkSemaphore <$> newEmptyMVar
+    ucUpdateSemaphore <- newEmptyMVar
+    lcLrcSync <- mkLrcSyncData >>= newTVarIO
+    ncDiscoveryContext <-
+        case npDiscovery npNetwork of
+            Left peers -> pure (DCStatic peers)
+            Right kadParams ->
+                DCKademlia <$> createKademliaInstance npBaseParams kadParams
+    ncSlottingVar <- mkSlottingVar npSystemStart
+    ncSlottingContext <-
+        case npUseNTP of
+            True  -> SCNtp <$> mkNtpSlottingVar
+            False -> pure SCSimple
+    let runSlottingAction =
+            flip
+                Ether.runReadersT
+                ( Tagged @SlottingVar ncSlottingVar
+                , Tagged @SlottingContextSum ncSlottingContext) .
+            runSlotsDataRedirect . runSlotsRedirect
+    ncUserSecret <- newTVarIO $ npUserSecret
+    ncBlockRetrievalQueue <- liftIO $ newTBQueueIO Const.blockRetrievalQueueSize
+    ncInvPropagationQueue <- liftIO $ newTBQueueIO Const.propagationQueueSize
+    ncRecoveryHeader <- liftIO newEmptyTMVarIO
+    ncProgressHeader <- liftIO newEmptyTMVarIO
+    ncShutdownFlag <- newTVarIO False
+    ncStartTime <- StartTime <$> liftIO Time.getCurrentTime
+    ncLastKnownHeader <- newTVarIO Nothing
+    ucMemState <- runSlottingAction newMemVar
+    ucDownloadingUpdates <- newTVarIO mempty
+    ncSscContext <- untag @ssc sscCreateNodeContext sscnp
+    -- TODO synchronize the NodeContext peers var with whatever system
+    -- populates it.
+    peersVar <- newTVarIO mempty
+    let ctx shutdownQueue =
+            NodeContext
+            { ncConnectedPeers = ConnectedPeers peersVar
+            , ncLrcContext = LrcContext {..}
+            , ncUpdateContext = UpdateContext {..}
+            , ncNodeParams = np
+            , ncGenesisLeaders = genesisLeaders npCustomUtxo
+            , ncSendLock = Nothing
+            , ncShutdownNotifyQueue = shutdownQueue
+#ifdef WITH_EXPLORER
+            , ncTxpGlobalSettings = explorerTxpGlobalSettings
+#else
+            , ncTxpGlobalSettings = txpGlobalSettings
+#endif
+            , ..
+            }
+    -- This queue won't be used.
+    fakeQueue <- liftIO (newTBQueueIO 100500)
```

Comment:
`allWorkersCount` is terrible. Note that it was became less broken as a result of this change (it makes it consider slotting and DHT workers properly).